obspy/clients/fdsn/mass_downloader/utils.py

```diff
@@ -0,0 +1,534 @@
+    channel_tag = "{%s}Channel" % ns
+    response_tag = "{%s}Response" % ns
+
+    context = etree.iterparse(filename, events=("start", ),
+                              tag=(network_tag, station_tag, channel_tag,
+                                   response_tag))
+
+    channels = []
+    for event, elem in context:
+        if elem.tag == channel_tag:
+            channel = elem.get('code')
+            location = elem.get('locationCode').strip()
+            starttime = obspy.UTCDateTime(elem.get('startDate'))
+            end_date = elem.get('endDate')
+            if end_date is not None:
+                endtime = obspy.UTCDateTime(elem.get('endDate'))
+            else:
+                endtime = obspy.UTCDateTime(2599, 1, 1)
+        elif elem.tag == response_tag:
+            channels.append(ChannelAvailability(
+                network, station, location, channel, starttime, endtime,
+                filename))
+        elif elem.tag == station_tag:
+            station = elem.get('code')
+            location, channel, starttime, endtime = \
+                None, None, None, None
+        elif elem.tag == network_tag:
+            network = elem.get('code')
+            station, location, channel, starttime, endtime = \
+                None, None, None, None, None
+        while elem.getprevious() is not None:
+            del elem.getparent()[0]
+
+    return channels
+
+
+def get_stationxml_filename(str_or_fct, network, station, channels,
+                            starttime, endtime):
+    """
+    Helper function getting the filename of a StationXML file.
+
+    :param str_or_fct: The string or function to be evaluated.
+    :type str_or_fct: function or str
+    :param network: The network code.
+    :type network: str
+    :param station: The station code.
+    :type station: str
+    :param channels: The channels. Each channel is a tuple of two strings:
+        location code and channel code.
+    :type channels: list of tuples
+    :param starttime: The start time.
+    :type starttime: :class:`~obspy.core.utcdatetime.UTCDateTime`
+    :param endtime: The end time.
+    :type endtime: :class:`~obspy.core.utcdatetime.UTCDateTime`
+
+    The rules are simple, if it is a function, network, station, channels,
+    starttime, and endtime are passed as arguments.
+
+    If it is a string, and it contains ``"{network}"``, and ``"{station}"``
+    formatting specifiers, ``str.format()`` is called.
+
+    Otherwise it is considered to be a folder name and the resulting
+    filename will be ``"FOLDER_NAME/NET.STA.xml"``
+
+    This function will either return a string or a dictionary with three
+    keys: ``"available_channels"``, ``"missing_channels"``, and ``"filename"``.
+    """
+    # Call if possible.
+    if callable(str_or_fct):
+        path = str_or_fct(network, station, channels, starttime, endtime)
+    # Check if its a format template.
+    elif ("{network}" in str_or_fct) and ("{station}" in str_or_fct):
```

Comment:
Isn't this a bit strict? One could try to apply a formatting option to the substitutions.